File: src/lib/Transformation2D.cpp
#ifndef _Transformation2D
#define _Transformation2D

#include <cmath>
#include <cstring>
#include "../Constants.cpp"

struct Transformation2D {
	float data[6];

	// apply offset and basis vectors.
	void apply(float* points, int beg, int end, int stride) {
		for(int i=beg;i<end;i+=stride) {
			const float x = points[i+0];
			const float y = points[i+1];
			points[i+0] = x*data[0] + y*data[2] + data[4];
			points[i+1] = x*data[1] + y*data[3] + data[5];
		}
	}

	// only apply the offset part of this transformation.
	void applyOffset(float* points, int beg, int end, int stride) {
		for(int i=beg;i<end;i+=stride) {
			const float x = points[i+0];
			const float y = points[i+1];
			points[i+0] = data[4];
			points[i+1] = data[5];
		}
	}

	// only apply the basis part of this transformation.
	void applyBasis(float* points, int beg, int end, int stride) {
		for(int i=beg;i<end;i+=stride) {
			const float x = points[i+0];
			const float y = points[i+1];
			points[i+0] = x*data[0] + y*data[2];
			points[i+1] = x*data[1] + y*data[3];
		}
	}

	Transformation2D() {
		// xbasis
		this->data[0] = 1;
		this->data[1] = 0;
		// ybasis
		this->data[2] = 0;
		this->data[3] = 1;
		// offset
		this->data[4] = 0;
		this->data[5] = 0;
	}
	Transformation2D(float x, float y, float w, float h, float r) {
		const float s = std::sinf(r * PI_2);
		const float c = std::cosf(r * PI_2);
		// xbasis
		this->data[0] = c*w;
		this->data[1] = s*w;
		// ybasis
		this->data[2] = -s*h;
		this->data[3] =  c*h;
		// offset
		this->data[4] = x;
		this->data[5] = y;
	}
	Transformation2D(Transformation2D& a, Transformation2D& b) {
		memcpy(this->data, b.data, 6 * sizeof(float));
		a.applyBasis (this->data, 0, 6, 2);// apply basis to entire matrix.
		a.applyOffset(this->data, 4, 6, 2);// only apply offset to offset.
	}

	Transformation2D compose(Transformation2D b) {
		return Transformation2D(*this, b);
	}
};

#endif

File: src/application/lib/Transformation2D.js

class Transformation2D extends Array {
	constructor(...args) {
		if(args.length === 1) super(...args[0]);
		if(args.length === 0) super(0,0,  1,0,  0,1);
		if(args.length === 5) {
			super(6);
			const [x,y,w,h,r] = args;
			const s = Math.sin(r * 2.0 * Math.PI);
			const c = Math.cos(r * 2.0 * Math.PI);
			this[0] = x;	this[1] = y;
			this[2] =  c*w;	this[3] = s*w;
			this[4] = -s*h;	this[5] = c*h;
		}
	}
	
	// ============================================================
	// Composition
	// ------------------------------------------------------------
	
	compose(tran) {
		const lt  = this;
		const rt  = tran;
		const dst = new Transformation2D(rt);
		lt.applyBasis(dst, 0, 6, 2);
		dst[0] += lt[0];
		dst[1] += lt[1];
		return dst;
	}
	
	// ============================================================
	// Apply transformations to position data
	// ------------------------------------------------------------
	
	// apply offset and basis vectors.
	apply(points, beg, end, stride) {
		for(let i=beg;i<end;i+=stride) {
			const x = points[i+0];
			const y = points[i+1];
			points[i+0] = this[0] + x*this[2] + y*this[4];
			points[i+1] = this[1] + x*this[3] + y*this[5];
		}
		return points;
	}
	
	// only apply the offset part of this transformation.
	applyOffset(points, beg, end, stride) {
		for(let i=beg;i<end;i+=stride) {
			points[i+0] += this[0];
			points[i+1] += this[1];
		}
		return points;
	}
	
	// only apply the basis part of this transformation.
	applyBasis(points, beg, end, stride) {
		for(let i=beg;i<end;i+=stride) {
			const x = points[i+0];
			const y = points[i+1];
			points[i+0] = x*this[2] + y*this[4];
			points[i+1] = x*this[3] + y*this[5];
		}
		return points;
	}
	
};



